node/src/node.rs

```diff
@@ -498,6 +500,99 @@ where
             Err(e) => Err(e.into()),
         }
     }
+
+    /// Return a stream which will yield all the headers, as they are being received by the node,
+    /// starting from the first header received after the call. Stream is guaranteed to return either
+    /// header or error for each height, in order.
+    pub async fn header_subscribe(
+        &self,
+    ) -> Result<ReceiverStream<Result<ExtendedHeader, SubscriptionError>>> {
+        let store = self
+            .store
+            .as_ref()
+            .cloned()
+            .expect("store should be present");
+        let mut prev_head = store.head_height().await?;
+
+        let (tx, rx) = mpsc::channel(16);
+
+        spawn(async move {
+            loop {
+                let new_head = store.wait_new_head().await;
+                for height in (prev_head + 1)..=new_head {
+                    let header_or_error =
+                        store
+                            .get_by_height(height)
+                            .await
+                            .map_err(|e| SubscriptionError {
+                                height,
+                                source: NodeError::Store(e),
+                            });
+                    if tx.send(header_or_error).await.is_err() {
+                        return; // receiver dropped
+                    }
+                    prev_head = new_head;
+                }
+            }
+        });
+
+        Ok(ReceiverStream::new(rx))
+    }
+
+    /// Return a stream which will yield all the blobs from the namespace, as the new headers
+    /// are being received by the node starting from the first header received after the call.
+    /// Stream is guaranteed to return all blobs (possibly zero) or error for each height, in order.
+    pub async fn blob_subscribe(
+        &self,
+        namespace: Namespace,
+    ) -> Result<ReceiverStream<Result<(u64, Vec<Blob>), SubscriptionError>>> {
+        let store = self
+            .store
+            .as_ref()
+            .cloned()
+            .expect("store should be present");
+        let p2p = self.p2p.as_ref().cloned().expect("p2p should be present");
+
+        let mut prev_head = store.head_height().await?;
```

Comment:
This is wrong in the following edge case:

- Let's say I used lumina node 300 blocks ago and then terminated it.
- I start a new node and I immediately call `header_subscribe`
- I will start getting very old headers because of very old `head_height`

I believe the `prev_head` needs to be initialized based on `wait_new_head`.

For example something like this:

```rust
        spawn(async move {
            let mut prev_head = None;

            loop {
                let new_head = store.wait_new_head().await;

                if prev_head.is_none() {
                    prev_head = Some(new_head - 1);
                }

                for height in prev_head.unwrap() + 1..=new_head {
             // ....
```

Reply:
Moved the source of truth about subscription pacing to syncer. `HeightSequencer` is now responsible for publishing heights of new headers received by the node, guaranteeeing that they are in order (potentially waiting for syncer to fill in the gap in case we miss few a and need to fill back). It also waits for the head init, so this should avoid mentioned issue.